### PR TITLE
Speed up the `Tap#formula_files` and `Tap#cask_files` methods

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

These are slow because of some Pathname related reasons so it's been changed up to use `Dir.glob` instead and that makes a large difference. This is an alternative approach to #18163 that is a little bit faster.

```
$ hyperfine --parameter-list branch master,speed-up-loading-cask-and-formula-file-names,speed-up-loading-cask-and-formula-file-names_v2 --warmup 5 --setup 'git switch {branch}' 'brew cat gimp'
Benchmark 1: brew cat gimp (branch = master)
  Time (mean ± σ):      2.150 s ±  0.014 s    [User: 1.369 s, System: 0.745 s]
  Range (min … max):    2.126 s …  2.174 s    10 runs

Benchmark 2: brew cat gimp (branch = speed-up-loading-cask-and-formula-file-names)
  Time (mean ± σ):      1.814 s ±  0.012 s    [User: 1.068 s, System: 0.711 s]
  Range (min … max):    1.797 s …  1.840 s    10 runs

Benchmark 3: brew cat gimp (branch = speed-up-loading-cask-and-formula-file-names_v2)
  Time (mean ± σ):      1.489 s ±  0.009 s    [User: 0.905 s, System: 0.550 s]
  Range (min … max):    1.478 s …  1.503 s    10 runs

Summary
  brew cat gimp (branch = speed-up-loading-cask-and-formula-file-names_v2) ran
    1.22 ± 0.01 times faster than brew cat gimp (branch = speed-up-loading-cask-and-formula-file-names)
    1.44 ± 0.01 times faster than brew cat gimp (branch = master)
```

### CODE REVIEW

**Review by Bo98 — APPROVED** [review]
Wow, didn't expect `glob` to be that much faster. Maybe we need to come up with some alternative to Pathname at this point.

File: Library/Homebrew/tap.rb [hunk 1]
```diff
@@ -712,10 +712,10 @@ def formula_files
       if formula_dir == path
         # We only want the top level here so we don't treat commands & casks as formulae.
         # Sharding is only supported in Formula/ and HomebrewFormula/.
-        formula_dir.children
+        Dir.glob(File.join(formula_dir.to_s, "*.rb"))
       else
-        formula_dir.find
-      end.select { formula_file?(_1) }
+        Dir.glob(File.join(formula_dir.to_s, "**/*.rb"))
```
Comment (carlocab) [inline Library/Homebrew/tap.rb:717]:
```suggestion
        formula_dir.glob("**/*.rb")
```
Is this also a lot slower?

Comment (apainintheneck) [reply Library/Homebrew/tap.rb:717]: Yeah, that's still significantly slower.

```diff
diff --git a/Library/Homebrew/tap.rb b/Library/Homebrew/tap.rb
index 0fa277417d..39332b9319 100644
--- a/Library/Homebrew/tap.rb
+++ b/Library/Homebrew/tap.rb
@@ -712,10 +712,10 @@ class Tap
       if formula_dir == path
         # We only want the top level here so we don't treat commands & casks as formulae.
         # Sharding is only supported in Formula/ and HomebrewFormula/.
-        Dir.glob(File.join(formula_dir.to_s, "*.rb"))
+        formula_dir.glob("*.rb")
       else
-        Dir.glob(File.join(formula_dir.to_s, "**/*.rb"))
-      end.map { Pathname(_1) }
+        formula_dir.glob("**/*.rb")
+      end
     else
       []
     end
@@ -736,7 +736,7 @@ class Tap
   sig { returns(T::Array[Pathname]) }
   def cask_files
     @cask_files ||= if cask_dir.directory?
-      Dir.glob(File.join(cask_dir.to_s, "**/*.rb")).map { Pathname(_1) }
+      cask_dir.glob("**/*.rb")
     else
       []
     end
```

```console
$ hyperfine --parameter-list branch speed-up-loading-cask-and-formula-file-names_v2,speed-up-loading-cask-and-formula-file-names_v3 --warmup 5 --setup 'git switch {branch}' 'brew cat gimp'
Benchmark 1: brew cat gimp (branch = speed-up-loading-cask-and-formula-file-names_v2)
  Time (mean ± σ):      1.499 s ±  0.009 s    [User: 0.909 s, System: 0.553 s]
  Range (min … max):    1.488 s …  1.521 s    10 runs

Benchmark 2: brew cat gimp (branch = speed-up-loading-cask-and-formula-file-names_v3)
  Time (mean ± σ):      1.721 s ±  0.010 s    [User: 1.134 s, System: 0.552 s]
  Range (min … max):    1.707 s …  1.737 s    10 runs

Summary
  brew cat gimp (branch = speed-up-loading-cask-and-formula-file-names_v2) ran
    1.15 ± 0.01 times faster than brew cat gimp (branch = speed-up-loading-cask-and-formula-file-names_v3)
```

**Review by carlocab — APPROVED** [review]

**Comment by apainintheneck** [comment]
> Wow, didn't expect `glob` to be that much faster. Maybe we need to come up with some alternative to Pathname at this point.

As you already know, doing a bunch of `Pathname` operations in a loop can be slow. We don't need to do that as often anymore because of the API which is good so I'm not sure it'd really worth it to create our own wrapper class. This is just a really bad example. It does make me wonder how some of the methods on the `Pathname` class are implemented though since it they seem to be so much slower than the equivalent `File` methods at times.

**Comment by Bo98** [comment]
> Edit: It looks like most of `Pathname` is implemented in C which makes me wonder even more. Could it just be that we have a lot of `Pathname` extensions that are slowing things down a lot?

It's not entirely C - most/all of the slow bits are in Ruby, particularly the ones that use regexes.

**Comment by apainintheneck** [comment]
You're probably right. I did see our old friend `Pathname#relative_path_from`, which is notoriously slow, is implemented in Ruby. Interestingly enough `Pathname#glob` is implemented in C but I think it might call some methods implemented in Ruby.

**Comment by Bo98** [comment]
Yeah `Pathname#glob` calls `Pathname#+` which is slow.

`Pathname.glob` is different and may be roughtly equivelent to what we have now.

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks again @apainintheneck!

**Pull request merged**